core/kernel/tee_ta_manager.c

```diff
@@ -463,6 +481,25 @@ static TEE_Result tee_ta_init_session_with_context(struct tee_ta_ctx *ctx,
 	return TEE_SUCCESS;
 }
 
+static uint32_t new_session_id(struct tee_ta_session_head *open_sessions)
+{
+	struct tee_ta_session *s = NULL;
+	uint32_t id = 1;
+
+again:
```

Comment:
I've been playing with this part in GDB, and for entire xtest, `id` never gets bigger than 6 and it is unlikely that it will ever become that big. But if it would be a bigger number, then the loop below is pretty in-efficient, since it restarts with `s->id = 1` every time and counts up to the `id` that we just incremented. But, I probably misunderstand some implementation details here.

Reply:
Session ID 1 might have been freed since the last call to `new_session_id()`.

Reply:
We could start from the ID of the last session +1 in the list and allow wrapping but bail out if we're back at the ID where we started.

Reply:
The problem is where to store the ID of the last session. There are multiple session lists. One for sessions opened from normal world (1 below) but also each TA has its own session list for TA to TA (2 below). A session list is just a bunch of linked `struct tee_ta_session` objects, so no place to store the last ID.
```
Cscope tag: tee_ta_open_session
   #   line  filename / context / line
   1    313  /home/jerome/work/optee_build_qemu/optee_os/core/arch/arm/tee/entry_std.c <<entry_open_session>>
             res = tee_ta_open_session(&err_orig, &s, &tee_open_sessions, &uuid,
   2    781  /home/jerome/work/optee_build_qemu/optee_os/core/tee/tee_svc.c <<syscall_open_ta_session>>
             res = tee_ta_open_session(&ret_o, &s, &utc->open_sessions, uuid,
```

Reply:
The IDs are local to the head structure open_sessions, so TAILQ_LAST on that should be enough. The alternative is to replace the open_sessions structure with a struct handle_db. We'd need to extend the handle interface to allow iterating or at least a callback to be used on each element when destroying the database.

Reply:
On the other hand, this is something we can optimize later if needed.

Reply:
I have an update ready, please see the "[Fixup]" commit.

Reply:
How about using the function `tee_ta_find_session_nolock()` to see if the ID already is taken and then have a conventional loop instead of a goto-based to iterate of the IDs?